ui/packages/catalog-ui-search/src/main/webapp/js/controllers/cesium.layerCollection.controller.js

```diff
@@ -100,37 +108,50 @@ var Controller = CommonLayerController.extend({
     }
   },
   setAlpha: function(model) {
-    var layer = this.layerForCid[model.id]
+    const layer = this.layerForCid[model.id]
     layer.alpha = model.get('alpha')
   },
   setShow: function(model) {
-    var layer = this.layerForCid[model.id]
+    if (!this.layerForCid[model.id]) {
+      this.initLayer(model)
+    }
+    const layer = this.layerForCid[model.id]
     layer.show = model.shouldShowLayer()
   },
   /*
-            removing/re-adding the layers causes visible "re-render" of entire map;
-            raising/lowering is smoother.
-            raising means to move to a higher index.  higher indexes are displayed on top of lower indexes.
-            so we have to reverse the order property here to make it display correctly.  
-            in other words, order 1 means highest index.
-        */
+    removing/re-adding the layers causes visible "re-render" of entire map;
+    raising/lowering is smoother.
+    raising means to move to a higher index.  higher indexes are displayed on top of lower indexes.
+    so we have to reverse the order property here to make it display correctly.  
+    in other words, order 1 means highest index.
+  */
   reIndexLayers: function() {
     this.collection.forEach(function(model, index) {
-      var layer = this.layerForCid[model.id]
-      var previousOrder = this.map.imageryLayers.indexOf(layer) + 1
-      var currentOrder = this.collection.length - model.get('order') // order is backwards on cesium (higher indexes are displayed above lower)
-      var method = currentOrder > previousOrder ? 'raise' : 'lower' // raise means move to higher index :(
-      var count = Math.abs(currentOrder - previousOrder)
-      // console.log(method + " " + model.get('name') + " " + count);  // useful for debugging!
-      _.times(
-        count,
-        function() {
-          this.map.imageryLayers[method](layer)
-        },
-        this
-      )
+      if (this.layerForCid[model.id]) {
+        const layer = this.layerForCid[model.id]
+        const previousOrder = this.map.imageryLayers.indexOf(layer)
+        const currentOrder = this.getLayerIndex(model)
+        const method = currentOrder > previousOrder ? 'raise' : 'lower' // raise means move to higher index :(
+        const count = Math.abs(currentOrder - previousOrder)
+        // console.log(method + " " + model.get('name') + " " + count);  // useful for debugging!
+        _.times(
+          count,
+          function() {
+            this.map.imageryLayers[method](layer)
+          },
+          this
+        )
+      }
     }, this)
   },
+  getLayerIndex: function(model) {
```

Comment:
This hurts my head, but so did the old code 😅 

The layer ordering logic is really easy to get wrong, so we should thoroughly test these changes with a good number of layers and verify that things end up in the right order when we play around with ordering.

I wonder if we could write a few unit tests around this area to ensure functionality in the future.

Reply:
Yeah, originally I tested with only 3 layers. @jrnorth found an issue with both my initialization and reindexing logic. For some reason, the reindexing logic worked as is, but I changed it anyways to make it easier to reason about. Tested it with 5 layers and it seems alright for now. Could use some pointers for writing unit tests 😅 Would I just make functions that just output where the layer should be inserted?

Reply:
I second Andrew's suggestion, layer ordering needs to be unit tested.